Simplify quantified expressions over constants

tautschnig opened the pull request:
Other simplification rules may turn the where-clause of a quantified expression into a (Boolean) constant. We can completely remove such quantified expressions so as not to distract back-ends that may special-case quantified expressions.

We saw such expressions when working on mlkem-native.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Simplify quantified expressions over constants

Other simplification rules may turn the where-clause of a quantified
expression into a (Boolean) constant. We can completely remove such
quantified expressions so as not to distract back-ends that may
special-case quantified expressions.

We saw such expressions when working on mlkem-native.

kroening commented:
It's unfortunately not that easy: The domain may be empty, and then ∀x.false is true, and ∃x.true is false.

tautschnig commented:
> It's unfortunately not that easy: The domain may be empty, and then ∀x.false is true, and ∃x.true is false.

Now fixed, and fixed the implementation in the propositional back-end as well.

rod-chapman commented:
Can we get a code-owner review on this one please? I will try it out on mlkem-native/mldsa-native and post results here.

rod-chapman commented:
OK. Proof results on mlkem-native and mldsa-native both look good. 100% SUCCESS with no significant change in performance.
